flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java

```diff
@@ -0,0 +1,259 @@
+				throw new FlinkRuntimeException("Failed to assign splits due to", e);
+			}
+		}
+
+		// Ensure all the subtasks in the assignment have registered.
+		for (Integer subtaskId : assignment.assignment().keySet()) {
+			if (!registeredReaders.containsKey(subtaskId)) {
+				throw new IllegalArgumentException(String.format(
+						"Cannot assign splits %s to subtask %d because the subtask is not registered.",
+						registeredReaders.get(subtaskId), subtaskId));
+			}
+		}
+
+		assignmentTracker.recordSplitAssignment(assignment);
+		assignment.assignment().forEach(
+				(id, splits) -> {
+					try {
+						operatorCoordinatorContext.sendEvent(new AddSplitEvent<>(splits), id);
+					} catch (TaskNotRunningException e) {
+						throw new FlinkRuntimeException(String.format(
+								"Failed to assign splits %s to reader %d.", splits, id), e);
+					}
+				});
+	}
+
+	@Override
+	public <T> void callAsync(
+			Callable<T> callable,
+			BiConsumer<T, Throwable> handler,
+			long initialDelay,
+			long period) {
+		notifier.notifyReadyAsync(callable, handler, initialDelay, period);
+	}
+
+	@Override
+	public <T> void callAsync(Callable<T> callable, BiConsumer<T, Throwable> handler) {
+		notifier.notifyReadyAsync(callable, handler);
+	}
+
+	// --------- Package private additional methods for the SourceCoordinator ------------
+
+	/**
+	 * Take a snapshot of this SourceCoordinatorContext.
+	 *
+	 * @param checkpointId The id of the ongoing checkpoint.
+	 * @param splitSerializer The serializer of the splits.
+	 * @param out An ObjectOutput that can be used to
+	 */
+	void snapshotState(
+			long checkpointId,
+			SimpleVersionedSerializer<SplitT> splitSerializer,
+			ObjectOutput out) throws Exception {
+		out.writeObject(registeredReaders);
+		assignmentTracker.snapshotState(checkpointId, splitSerializer, out);
+	}
+
+	/**
+	 * Restore the state of the context.
+	 * @param splitSerializer the serializer for the SourceSplits.
+	 * @param in the input from which the states are read.
+	 * @throws Exception when the restoration failed.
+	 */
+	@SuppressWarnings("unchecked")
+	void restoreState(
+			SimpleVersionedSerializer<SplitT> splitSerializer,
+			ObjectInput in) throws Exception {
+		Map<Integer, ReaderInfo> readers = (Map<Integer, ReaderInfo>) in.readObject();
+		registeredReaders.putAll(readers);
+		assignmentTracker.restoreState(splitSerializer, in);
+	}
+
+	/**
+	 * Register a source reader.
+	 *
+	 * @param readerInfo the reader information of the source reader.
+	 */
+	void registerSourceReader(ReaderInfo readerInfo) {
+		registeredReaders.put(readerInfo.getSubtaskId(), readerInfo);
+	}
+
+	/**
+	 * Unregister a source reader.
+	 *
+	 * @param subtaskId the subtask id of the source reader.
+	 */
+	void unregisterSourceReader(int subtaskId) {
+		Preconditions.checkNotNull(registeredReaders.remove(subtaskId),
+				"Failed to unregister source reader of id " + subtaskId + " because it is not registered.");
```

Comment:
Avoid string concatenation in Preconditions, because they always eagerly happen, even on the "happy path". You can use `%s` as a placeholder, similar to `{}` in log statements.